minerl/env/observations.py

```diff
@@ -0,0 +1,116 @@
+        return None
+    item = []
+    block = []
+    with open(os.path.join(dir_path, 'data', 'items.txt')) as f:
+        for line in f:
+            item.append(line.strip())
+
+    with open(os.path.join(dir_path, 'data', 'blocks.txt')) as f:
+        for line in f:
+            block.append(line.strip())
+    return item, block
+
+
+ITEMS, BLOCKS = get_item_block_lists()
+# Use these for observation spaces that interact with minecraft items and blocks.
+#   Provides a non-meaningful but consistent ordering over the space of names.
+# Currently no support for entities.
+
+
+def pov_observation(info, obs_space):
+    """
+    :param info: Info dict generated by Malmo mod. Must contain 'pov'.
+    :param obs_space: unused.
+    :return: np.array with shape specified in Mission xml file.
+    """
+    return info['pov']
+
+
+def inventory_observation(info, obs_space):
+    """
+    For every item type in the observation space, add up how many of said item are in the inventory.
+    Duplicates the MineRL default inventory observation.
+
+    :param info: Info dict generated by Malmo mod.
+    :param obs_space: The observation space. Used to decide which item information to include.
+    :return: Dict of {item name : count in inventory}.
+    """
+    inventory_spaces = obs_space['inventory'].spaces
+
+    inventory_dict = {k: 0 for k in inventory_spaces}
+    # TODO change to maalmo
+    if 'inventory' in info:
+        for stack in info['inventory']:
+            if 'type' in stack and 'quantity' in stack:
+                type_name = stack['type']
+                if type_name == 'log2':
+                    type_name = 'log'
+
+                try:
+                    inventory_dict[type_name] += stack['quantity']
+                except ValueError:
+                    continue
+                except KeyError:
+                    # We only care to observe what was specified in the space.
+                    continue
+    else:
+        print("Inventory information could not be found, returning empty inventory.")
+
+    return inventory_dict
+
+
+def observation_creator(observation_space, info, extra_handlers=None):
+    """
+    Parses through the requested observation spaces and attempts to create each requested observation
+        from handlers defined in this file, using the given info dict.
+    
+    :param observation_space: Gym/MineRL observation space.
+    :param info: Python dict.
+    :param extra_handlers: Any extra functions to be used in decoding the observation space.
+    :return: Python dict with one index corresponding to each index in the observation space that has
+        an associated handler.
+    """
+    obs_space = deepcopy(observation_space.spaces)
+    obs_dict = {}
+
+    handlers = {
+        'pov': pov_observation,
+        'inventory': inventory_observation,
+    }
+    if extra_handlers is not None:
+        handlers.update(extra_handlers)
+
+    for key, handler_fn in handlers.items():
```

Comment:
My intuition here would be that it'd make more sense to go through each element of the observation space and find a handler for it, rather than going through all of the handlers and skipping the ones that aren't in the obs space (I think it does the same thing, this just feels like a less intuitive way to do it)

Reply:
This also has the benefit that you'd be able to provide a clearer error, because if an obs space key isn't found as you go through the loop, you can directly raise an error mentioning which obs space key it was that didn't have an associated handler key

Reply:
That's fair. Will do.